src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBaseKernels.hpp

```diff
@@ -129,128 +132,353 @@ struct MobilityKernel
+    m_density( fluid.density() ),
+    m_dDensity_dPres( fluid.dDensity_dPressure() ),
+    m_localMatrix( localMatrix ),
+    m_localRhs( localRhs )
+  {}
+
+  /**
+   * @struct StackVariables
+   * @brief Kernel variables (dof numbers, jacobian and residual) located on the stack
+   */
+  struct StackVariables
+  {
+public:
+
+    // Pore volume information
+
+    /// Pore volume at time n+1
+    real64 poreVolumeNew = 0.0;
+
+    /// Pore volume at the previous converged time step
+    real64 poreVolume_n = 0.0;
+
+    /// Derivative of pore volume with respect to pressure
+    real64 dPoreVolume_dPres = 0.0;
+
+    // Residual information
+
+    /// Index of the matrix row/column corresponding to the dof in this element
+    globalIndex dofNumber;
+
+    /// Storage for the element local residual vector
+    real64 localResidual;
+
+    /// Storage for the element local Jacobian matrix
+    real64 localJacobian;
+
+  };
+
+  /**
+   * @brief Getter for the ghost rank of an element
+   * @param[in] ei the element index
+   * @return the ghost rank of the element
+   */
+  GEOSX_HOST_DEVICE
+  integer elemGhostRank( localIndex const ei ) const
+  { return m_elemGhostRank( ei ); }
+
+
+  /**
+   * @brief Performs the setup phase for the kernel.
+   * @param[in] ei the element index
+   * @param[in] stack the stack variables
+   */
+  GEOSX_HOST_DEVICE
+  void setup( localIndex const ei,
+              StackVariables & stack ) const
+  {
+    // initialize the pore volume
+    stack.poreVolumeNew = ( m_volume[ei] + m_deltaVolume[ei] ) * m_porosityNew[ei][0];
+    stack.poreVolume_n = m_volume[ei] * m_porosity_n[ei][0];
+    stack.dPoreVolume_dPres = ( m_volume[ei] + m_deltaVolume[ei] ) * m_dPoro_dPres[ei][0];
+
+    // set degree of freedom index for this element
+    stack.dofNumber = m_dofNumber[ei];
+  }
+
+  /**
+   * @brief Compute the local accumulation contributions to the residual and Jacobian
+   * @tparam FUNC the type of the function that can be used to customize the kernel
+   * @param[in] ei the element index
+   * @param[inout] stack the stack variables
+   * @param[in] kernelOp the function used to customize the kernel
+   */
+  template< typename FUNC = NoOpFunc >
+  GEOSX_HOST_DEVICE
+  void computeAccumulation( localIndex const ei,
+                            StackVariables & stack,
+                            FUNC && kernelOp = NoOpFunc{} ) const
   {
     // Residual contribution is mass conservation in the cell
-    localAccum = poreVolNew * densNew - poreVol_n * dens_n;
+    stack.localResidual = stack.poreVolumeNew * m_density[ei][0] - stack.poreVolume_n * m_density_n[ei][0];
 
     // Derivative of residual wrt to pressure in the cell
-    localAccumJacobian = dPoreVol_dPres * densNew + dDens_dPres * poreVolNew;
+    stack.localJacobian = stack.dPoreVolume_dPres * m_density[ei][0] + m_dDensity_dPres[ei][0] * stack.poreVolumeNew;
+
+    // Customize the kernel with this lambda
+    kernelOp();
```

Comment:
we should be able to use this to add the thermal derivatives so I would say that this design is perfect.